tools/xenpcid/xenpcid.c

```diff
@@ -0,0 +1,1056 @@
+        dp = (char *)pcid_zalloc(PATH_MAX);
+        if ( !(realpath(spath, dp)) ) {
+            fprintf(stderr, "realpath() failed\n");
+            goto fail;
+        }
+        *result = dp;
+        new_path = (char *)pcid_zalloc(strlen(dp) + strlen("/unbind") + 1);
+        /* Unbind from the old driver */
+        sprintf(new_path, "%s/unbind", dp);
+
+        if ( handle_write_cmd(new_path, pci_info) != 0 ) {
+            fprintf(stderr, "Couldn't unbind device\n");
+            goto fail_write;
+        }
+        free(new_path);
+    } else {
+        *result = (char *)pcid_zalloc(strlen("nolstat") + 1);
+        sprintf(*result, "nolstat");
+    }
+    free(spath);
+    free(pci_info);
+
+    return 0;
+
+fail_write:
+    free(new_path);
+fail:
+    free(spath);
+    free(pci_info);
+
+    return -1;
+}
+
+static int handle_reset_cmd(char *pci_path, char *pci_info)
+{
+    char *reset;
+    int rc, fd;
+
+    reset = (char *)pcid_zalloc(strlen(SYSFS_PCIBACK_DRIVER) + strlen(pci_path) + 1);
+    sprintf(reset, SYSFS_PCIBACK_DRIVER"%s", pci_path);
+    free(pci_path);
+    fd = open(reset, O_WRONLY);
+    if (fd >= 0) {
+        rc = write(fd, pci_info, strlen(pci_info));
+        if (rc < 0)
+            fprintf(stderr, "write to %s returned %d\n", reset, rc);
+        close(fd);
+        free(reset);
+        return rc < 0 ? rc : 0;
+    }
+    if (errno != ENOENT)
+        fprintf(stderr, "Failed to access pciback path %s\n", reset);
+    free(reset);
+    reset = (char *)pcid_zalloc(strlen(SYSFS_PCI_DEV) + strlen(pci_info) +
+                                strlen("//reset") + 1);
+    sprintf(reset, "%s/%s/reset", SYSFS_PCI_DEV, pci_info);
+    fd = open(reset, O_WRONLY);
+    if (fd >= 0) {
+        rc = write(fd, "1", 1);
+        if (rc < 0)
+            fprintf(stderr, "write to %s returned %d\n", reset, rc);
+        close(fd);
+        free(reset);
+        return rc < 0 ? rc : 0;
+    }
+    if (errno == ENOENT) {
+        fprintf(stderr,
+                "The kernel doesn't support reset from sysfs for PCI device %s\n",
+                pci_info);
+    } else {
+        fprintf(stderr, "Failed to access reset path %s\n", reset);
+    }
+    free(reset);
+    return -1;
+}
+
+static int vchan_handle_message(libxl_ctx *ctx,
+                                struct vchan_state *state,
+                                struct pcid__json_object **resp)
+{
+    long long read_result;
+    int ret;
+    struct list_head *dir_list = NULL;
+    struct list_resources *resources_list = NULL;
+    struct pcid__json_object *command_obj, *args, *dir_id, *pci_info;
+    struct pcid__json_object *pci_path;
+    char *dir_name, *command_name, *full_path = NULL;
+
+    command_obj = pcid__json_map_get(XENPCID_MSG_EXECUTE, *resp, JSON_ANY);
+    command_name = command_obj->u.string;
+
+    if (strcmp(XENPCID_CMD_LIST, command_name) == 0) {
+        args = pcid__json_map_get(XENPCID_MSG_FIELD_ARGS, *resp, JSON_MAP);
```

Comment:
`pcid__json_map_get()` there and elsewhere below can return `NULL`. You need to check for this before trying using the value.

Reply:
Also, it will be move all this (and for other `if`branges)  logic  to `handle_*_function`.  

`vchan_handle_message` will be much easier to read after this.

Reply:
Done